<a href="https://colab.research.google.com/github/Mauricio-Fragoso-Mendoza/Simulacion-I/blob/main/C%C3%B3digo_2_Variables_Antit%C3%A9ticas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [20]:
import numpy as np
import random as r
import matplotlib.pyplot as plt
import scipy.stats as stats

Se estimará el resultado de la siguiente integral:
$$I=\int_0^1 e^{x^2}dx$$

Para este caso, se aplicará el Método de Montecarlo para variables antitéticas.

In [21]:
def f(x):
  return np.exp(x**2)

In [22]:
f(0)

np.float64(1.0)

In [23]:
f(1)

np.float64(2.718281828459045)

Se generarán variables $U \sim U(0,1)$, y se generarán 1000 datos.

In [24]:
def va1(f,N):
  s = 0
  for i in range (N):
      U = r.random()
      s += f(U)
  return s/N

In [35]:
l = []
for i in range (1000):
  M = va1(f,1000)
  l.append(M)
print(np.mean(l))
print(np.var(l))

1.4629325163792597
0.00022603134117789192


In [27]:
def va2(f,N):
  s = 0
  for i in range (N):
      U = r.random()
      s += f(1-U)
  return s/N

In [36]:
l = []
for i in range (1000):
  M = va2(f,1000)
  l.append(M)
print(np.mean(l))
print(np.var(l))

1.4623839021683196
0.0002225253184796888


In [30]:
def va(f,N):
  s = 0
  for i in range (N):
      U = r.random()
      s += (f(U)+f(1-U))/2
  return s/N

In [37]:
l = []
for i in range (1000):
  V = va(f,1000)
  l.append(M)
print(np.mean(l))
print(np.var(l))

1.450078014449409
1.9721522630525295e-31


In [34]:
Datos = l
n = len(Datos)
Media = np.mean(Datos)
Desviación_Estandar = np.std(Datos, ddof=1)

# Valor crítico (para 95% de confianza)
alpha = 0.05
valor_z = stats.t.ppf(1 - alpha / 2, df=n-1)

# Margen de error
Margen_Error = valor_z * (Desviación_Estandar / np.sqrt(n))

# Límites
Limite_Inferior = Media - Margen_Error
Limite_Superior = Media + Margen_Error

#Resultados
print("Media: ", Media)
print("Desviación Estándar: ", Desviación_Estandar)
print("Intervalo de confianza: (",Limite_Inferior,", ",Limite_Superior, ")")

Media:  1.4668254382246309
Desviación Estándar:  6.694896738718451e-16
Intervalo de confianza: ( 1.4668254382246306 ,  1.466825438224631 )


Por lo que el resultado de la integral es de aproximadamente: $I \approx 1.5$